database/sinkdb/state.go

```diff
@@ -41,25 +41,38 @@ func (s *state) SetAppliedIndex(index uint64) {
 	s.appliedIndex = index
 }
 
-// SetPeerAddr sets the address for the given peer.
-func (s *state) SetPeerAddr(id uint64, addr string) {
+// Peers returns the current set of peer nodes. The returned
+// map must not be modified.
+func (s *state) Peers() map[uint64]string {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	s.peers[id] = addr
+	return s.peers
 }
 
-// GetPeerAddr gets the current address for the given peer, if set.
-func (s *state) GetPeerAddr(id uint64) (addr string) {
+// SetPeerAddr sets the address for the given peer.
+func (s *state) SetPeerAddr(id uint64, addr string) {
+	newPeers := make(map[uint64]string)
```

Comment:
Since Go maps aren't safe for concurrent access, readers would need to keep s.mu locked while accessing the map. [This change](https://github.com/chain/chain/pull/1269/files#r119689333) would make that interface really clunky. By forcing all mutations of the map to create a new map, readers can call Peers and use the returned map without keeping a hold of s.mu. Changes to the peer list should be infrequent, so it seems like a reasonable tradeoff.